src/lib/device-state.ts

```diff
@@ -97,6 +97,30 @@ export const serviceInstallFromImage = (
 export const formatImageLocation = (imageLocation: string) =>
 	imageLocation.toLowerCase();
 
+export type DeviceConfigHook = (
+	configVars: Dictionary<string>,
+	osVersion: string,
+) => void;
+
+export interface HookRegistry {
+	deviceConfig: DeviceConfigHook[];
+}
+
+const STATE_ENDPOINT_HOOKS: HookRegistry = {
+	deviceConfig: [],
+};
+
+export function registerHook<
+	K extends keyof HookRegistry,
+	Fn extends typeof STATE_ENDPOINT_HOOKS[K][number]
+>(name: K, fn: Fn) {
+	STATE_ENDPOINT_HOOKS[name].push(fn);
+}
+
+function runHook(name: keyof HookRegistry, args: any[]) {
+	STATE_ENDPOINT_HOOKS[name].forEach(hook => hook.apply(null, args));
```

Comment:
```suggestion
	STATE_ENDPOINT_HOOKS[name].forEach(hook => hook(...args));
```

Reply:
This unfortunately doesn't work, even after adapting the signature as you suggested below.

Reply:
You can do
```suggestion
	STATE_ENDPOINT_HOOKS[name].forEach(hook => hook(...args as Parameters<typeof hook>));
```
instead and still be able to use `...` instead